File: lab2Gauss/gaussKotlin/app/src/main/java/com/example/gauss/GaussSolver.kt
package com.example.gauss

import kotlin.math.abs

class GaussSolver {
    private var sizeMatrix: Int = 0

    fun solve(A: List<List<Double>>, b: List<Double>): Result {
        val errorMessage = validateInput(A, b)
        if (errorMessage != "OK") {
            return Result.Failure(errorMessage)
        }

        val n = A.size
        sizeMatrix = n

        val matrix = A.map { it.toMutableList() }.toMutableList()
        val vector = b.toMutableList()
        val x = MutableList(n) { 0.0 }
        val IOR = MutableList(n) { it }

        for (k in 0 until n - 1) {
            val p = findPivot(matrix, IOR, k)
            if (p == -1) {
                return Result.Failure("Матрица вырождена!")
            }

            if (p != k) {
                IOR[k] = IOR[p].also { IOR[p] = IOR[k] }
            }

            val M = IOR[k]
            val AMAIN = matrix[M][k]

            if (abs(AMAIN) < 1e-15) {
                return Result.Failure("Нулевой ведущий элемент!")
            }

            for (j in k until n) {
                matrix[M][j] /= AMAIN
            }
            vector[M] /= AMAIN

            for (i in k + 1 until n) {
                val l = IOR[i]
                val factor = matrix[l][k]

                for (j in k + 1 until n) {
                    matrix[l][j] -= factor * matrix[M][j]
                }
                vector[l] -= factor * vector[M]
            }
        }

        val lastRow = IOR[n - 1]
        if (abs(matrix[lastRow][n - 1]) < 1e-15) {
            return Result.Failure("Система не имеет решения!")
        }

        x[n - 1] = vector[lastRow] / matrix[lastRow][n - 1]

        for (k in n - 2 downTo 0) {
            val row = IOR[k]
            var sum = 0.0

            for (j in k + 1 until n) {
                sum += matrix[row][j] * x[j]
            }
            x[k] = vector[row] - sum
        }

        return Result.Success(x)
    }

    private fun validateInput(A: List<List<Double>>, b: List<Double>): String {
        val n = A.size

        if (n == 0) {
            return "Матрица не может быть пустой!"
        }

        if (n > 20) {
            return "Слишком большая размерность (максимум 20)!"
        }

        for (i in 0 until n) {
            if (A[i].size != n) {
                return "Матрица должна быть квадратной!"
            }
        }

        if (b.size != n) {
            return "Размер вектора b должен совпадать с размером матрицы!"
        }

        for (i in 0 until n) {
            for (j in 0 until n) {
                if (A[i][j].isNaN() || A[i][j].isInfinite()) {
                    return "Матрица содержит некорректные значения!"
                }
            }
            if (b[i].isNaN() || b[i].isInfinite()) {
                return "Вектор b содержит некорректные значения!"
            }
        }

        return "OK"
    }

    private fun findPivot(A: List<List<Double>>, IOR: List<Int>, k: Int): Int {
        val n = A.size
        var maxVal = 0.0
        var pivotIndex = -1

        for (i in k until n) {
            val row = IOR[i]
            val currentVal = abs(A[row][k])

            if (currentVal > maxVal) {
                maxVal = currentVal
                pivotIndex = i
            }
        }

        return if (maxVal < 1e-15) -1 else pivotIndex
    }

    fun calculateResidual(A: List<List<Double>>, b: List<Double>, x: List<Double>): List<Double> {
        val n = A.size
        val residual = MutableList(n) { 0.0 }

        for (i in 0 until n) {
            var sum = 0.0
            for (j in 0 until n) {
                sum += A[i][j] * x[j]
            }
            residual[i] = sum - b[i]
        }

        return residual
    }

    fun calculateResidualNorm(residual: List<Double>): Double {
        return residual.maxOfOrNull { abs(it) } ?: 0.0
    }

    fun getMatrixSize(): Int = sizeMatrix

    sealed class Result {
        data class Success(val solution: List<Double>) : Result()
        data class Failure(val error: String) : Result()
    }
}

File: lab2Gauss/gauss/gaussSolver.cpp
#include "gaussSolver.h"
#include <cmath>
#include <algorithm>

GaussSolver::GaussSolver() : sizeMatrix(0) {}

bool GaussSolver::validateInput(const std::vector<std::vector<double>>& A,
                                const std::vector<double>& b,
                                std::string& errorMessage) {
    int n = A.size();

    if (n == 0) {
        errorMessage = "Матрица не может быть пустой!";
        return false;
    }

    if (n > 20) {
        errorMessage = "Слишком большая размерность (максимум 20)!";
        return false;
    }

    for (int i = 0; i < n; i++) {
        if (A[i].size() != n) {
            errorMessage = "Матрица должна быть квадратной!";
            return false;
        }
    }

    if (b.size() != n) {
        errorMessage = "Размер вектора b должен совпадать с размером матрицы!";
        return false;
    }

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (std::isnan(A[i][j]) || std::isinf(A[i][j])) {
                errorMessage = "Матрица содержит некорректные значения!";
                return false;
            }
        }
        if (std::isnan(b[i]) || std::isinf(b[i])) {
            errorMessage = "Вектор b содержит некорректные значения!";
            return false;
        }
    }

    errorMessage = "OK";
    return true;
}

int GaussSolver::findPivot(const std::vector<std::vector<double>> &A,
                           const std::vector<int>& IOR, int k) {
    int n = A.size();
    double maxVal = 0.0;
    int pivotIndex = -1;

    for (int i = k; i < n; i++) {
        int row = IOR[i];
        double currentVal = std::abs(A[row][k]);

        if (currentVal > maxVal) {
            maxVal = currentVal;
            pivotIndex = i;
        }
    }

    return (maxVal < 1e-15) ? -1 : pivotIndex;
}

bool GaussSolver::solve(std::vector<std::vector<double>> A,
                        std::vector<double> b,
                        std::vector<double> &x,
                        std::string& errorMessage) {

    if (!validateInput(A, b, errorMessage)) {
        return false;
    }

    int n = A.size();
    sizeMatrix = n;
    x.resize(n, 0.0);

    std::vector<int> IOR(n);
    for(int i = 0; i < n; i++) {
        IOR[i] = i;
    }

    for (int k = 0; k < n - 1; k++) {
        int p = findPivot(A, IOR, k);
        if (p == -1) {
            errorMessage = "Матрица вырождена!";
            return false;
        }

        if (p != k) {
            std::swap(IOR[k], IOR[p]);
        }

        int M = IOR[k];
        double AMAIN = A[M][k];

        if (std::abs(AMAIN) < 1e-15) {
            errorMessage = "Нулевой ведущий элемент!";
            return false;
        }

        for(int j = k; j < n; j++) {
            A[M][j] /= AMAIN;
        }
        b[M] /= AMAIN;

        for(int i = k + 1; i < n; i++) {
            int l = IOR[i];
            double factor = A[l][k];

            for(int j = k + 1; j < n; j++) {
                A[l][j] -= factor * A[M][j];
            }
            b[l] -= factor * b[M];
        }
    }

    int lastRow = IOR[n-1];
    if (std::abs(A[lastRow][n-1]) < 1e-15) {
        errorMessage = "Система не имеет решения!";
        return false;
    }

    x[n-1] = b[lastRow] / A[lastRow][n-1];

    for (int k = n-2; k >= 0; k--) {
        int row = IOR[k];
        double sum = 0.0;

        for(int j = k + 1; j < n; j++) {
            sum += A[row][j] * x[j];
        }
        x[k] = b[row] - sum;
    }

    errorMessage = "Решение найдено успешно!";
    return true;
}

std::vector<double> GaussSolver::calculateResidual(
    const std::vector<std::vector<double>>& A,
    const std::vector<double>& b,
    const std::vector<double>& x) {

    int n = A.size();
    std::vector<double> residual(n, 0.0);

    for (int i = 0; i < n; i++) {
        double sum = 0.0;
        for (int j = 0; j < n; j++) {
            sum += A[i][j] * x[j];
        }
        residual[i] = sum - b[i];
    }

    return residual;
}

double GaussSolver::calculateResidualNorm(const std::vector<double>& residual) {
    double norm = 0.0;
    for (double val : residual) {
        norm = std::max(norm, std::abs(val));
    }
    return norm;
}
